notebooks/02-ingestion/02-batch-ingestion.ipynb

```diff
@@ -0,0 +1,1008 @@
+    "WITH \"ext\" AS \n",
+    "(\n",
+    "    SELECT *\n",
+    "    FROM TABLE(\n",
+    "      EXTERN(\n",
+    "        '{\"type\":\"http\",\n",
+    "          \"uris\":[ \"https://druid.apache.org/data/wikipedia.json.gz\",\n",
+    "                   \"https://druid.apache.org/data/wikipedia.json.gz\",\n",
+    "                   \"https://druid.apache.org/data/wikipedia.json.gz\"\n",
+    "                 ]\n",
+    "         }',\n",
+    "        '{\"type\":\"json\"}'\n",
+    "      )\n",
+    ") EXTEND (\"isRobot\" VARCHAR, \"channel\" VARCHAR, \"timestamp\" VARCHAR, \"flags\" VARCHAR, \"isUnpatrolled\" VARCHAR, \"page\" VARCHAR, \"diffUrl\" VARCHAR, \"added\" BIGINT, \"comment\" VARCHAR, \"commentLength\" BIGINT, \"isNew\" VARCHAR, \"isMinor\" VARCHAR, \"delta\" BIGINT, \"isAnonymous\" VARCHAR, \"user\" VARCHAR, \"deltaBucket\" BIGINT, \"deleted\" BIGINT, \"namespace\" VARCHAR, \"cityName\" VARCHAR, \"countryName\" VARCHAR, \"regionIsoCode\" VARCHAR, \"metroCode\" BIGINT, \"countryIsoCode\" VARCHAR, \"regionName\" VARCHAR))\n",
+    "SELECT\n",
+    "  TIME_PARSE(\"timestamp\") AS \"__time\",\n",
+    "  *\n",
+    "FROM \"ext\"\n",
+    "PARTITIONED BY DAY\n",
+    "'''\n",
+    "druid.display.run_task(sql)\n",
+    "druid.sql.wait_until_ready('wikipedia_events_3')\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "41ce84c5-2c1c-4c6a-b9f4-7a22e2216ec5",
+   "metadata": {},
+   "source": [
+    "Let's look at the data now. The quantities are 3 times larger than before because we loaded the same file three times:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "f278c0d2-9fd5-4a7b-bb65-a6db7cf98971",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "druid.display.sql(\"\"\"\n",
+    "SELECT channel, count(*) num_events\n",
+    "FROM \"wikipedia_events_3\" \n",
+    "GROUP BY 1 \n",
+    "ORDER BY 2 DESC \n",
+    "LIMIT 10\n",
+    "\"\"\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "751316a2-bd0c-43eb-a74f-fed15782f54d",
+   "metadata": {},
+   "source": [
+    "#### Context Parameters\n",
+    "Certain aspects of the ingestion can be controlled using [context parameter](https://druid.apache.org/docs/latest/multi-stage-query/reference.html#context-parameters)s. This section discussed two of the commonly used parameters:\n",
+    "\n",
+    "##### maxNumTasks\n",
+    "The Multi-stage Query Framework uses parallel workers to execute each stage of the ingestion process. Each stage creates output partitions that organize the data in preparation for the next stage. \n",
+    "\n",
+    "The input stage parallelism is limited by the input sources, as each file is processed by one of the workers. While multiple input files are split evenly among parallel worker tasks. As such a single large file cannot be parallelized at this stage. Consider splitting single large files into multiple files to improve parallelism at this stage. \n",
```

Comment:
The final sentence is advisory, so not needed in a notebook.

Reply:
I don't agree. Replaced "Consider..." with "You can split very large files into smaller ones to improve parallelism for the input state.". But it is instructional and needed, people in the community have come up with things like "I am ingesting a single 40GB file and it is too slow..."